test/helper/change-detection.bash

```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+load "${REPOSITORY_ROOT}/test/helper/common"
+
+function wait_until_change_detection_event_begins() {
+  local MATCH_CONTENT='Change detected'
+  local MATCH_IN_LOG='/var/log/supervisor/changedetector.log'
+
+  _wait_until_expected_count_is_matched "${@}"
+}
+
+# NOTE: Change events can start and finish all within < 1 sec,
+# Reliably track the completion of a change event by counting events:
+function wait_until_change_detection_event_completes() {
+  local MATCH_CONTENT='Completed handling of detected change'
+  local MATCH_IN_LOG='/var/log/supervisor/changedetector.log'
+  
+  _wait_until_expected_count_is_matched "${@}"
+}
+
+function _get_logs_since_last_change_detection() {
+  local CONTAINER_NAME=$1
+  local MATCH_IN_FILE='/var/log/supervisor/changedetector.log'
+  local MATCH_STRING='Change detected'
+
+  # Read file in reverse, collect lines until match with sed is found,
+  # then stop and return these lines back in original order (flipped again through tac):
+  docker exec "${CONTAINER_NAME}" bash -c "tac ${MATCH_IN_FILE} | sed '/${MATCH_STRING}/q' | tac"
+}
```

Comment:
I had tried a few attempts before this, this works and seemed ok to go with?

---

I had come across the following which seemed to work, but if there was no match it returned the entire input:

```bash
docker exec "${CONTAINER_NAME}" bash -c "cat ${MATCH_IN_FILE} | sed -n 'H; /${MATCH_STRING}/h; \${g;p;}'"
```

Likewise, this seemed to work, but was starting from the first match found (instead of the last one) which wasn't desirable:

```bash
docker exec "${CONTAINER_NAME}" bash -c "cat ${MATCH_IN_FILE} | sed -n '/${MATCH_STRING}/,\$p'"
```

Reply:
I like the way it is at the moment best :)